src/Compilers/VisualBasic/Portable/Symbols/Tuples/TupleTypeSymbol.vb

```diff
@@ -0,0 +1,987 @@
+                                members.Add(New TupleEventSymbol(Me, DirectCast(member, EventSymbol).AsMember(currentUnderlying)))
+                            End If
+
+                        Case Else
+                            If currentNestingLevel = 0 Then
+                                Throw ExceptionUtilities.UnexpectedValue(member.Kind)
+                            End If
+                    End Select
+                Next
+
+                If currentUnderlying.Arity <> RestPosition Then
+                    Exit Do
+                End If
+
+                Dim oldUnderlying = currentUnderlying
+                currentUnderlying = oldUnderlying.TypeArgumentsNoUseSiteDiagnostics(RestPosition - 1).TupleUnderlyingType
+                currentNestingLevel += 1
+
+                If currentUnderlying.Arity <> RestPosition Then
+                    ' refresh members And target fields
+                    underlyingMembers = currentUnderlying.OriginalDefinition.GetMembers()
+                    currentFieldsForElements.Clear()
+                    CollectTargetTupleFields(currentUnderlying, currentFieldsForElements)
+                Else
+                    Debug.Assert(oldUnderlying.OriginalDefinition Is currentUnderlying.OriginalDefinition)
+                End If
+            Loop
+
+            currentFieldsForElements.Free()
+
+            ' At the end, add remaining virtual fields
+            For i As Integer = 0 To namesOfVirtualFields.Count - 1
+                Dim Name As String = namesOfVirtualFields(i)
+                If Name IsNot Nothing Then
+                    ' We couldn't find a backing field for this vitual field. It will be an error to access it. 
+                    Dim fieldRemainder As Integer ' one - based
+                    Dim fieldChainLength = NumberOfValueTuples(i + 1, fieldRemainder)
+                    Dim container As NamedTypeSymbol = GetNestedTupleUnderlyingType(_underlyingType, fieldChainLength - 1).OriginalDefinition
+
+                    Dim DiagnosticInfo = If(container.IsErrorType(),
+                                                          Nothing,
+                                                          ErrorFactory.ErrorInfo(ERRID.ERR_MissingRuntimeHelper,
+                                                                               container.Name & "." & TupleMemberName(fieldRemainder)))
+
+                    Dim defaultName As String = TupleMemberName(i + 1)
+                    ' Add a field with default name if the given name Is different
+                    If Name <> defaultName Then
+                        members.Add(New TupleErrorFieldSymbol(Me, defaultName, -members.Count - 1, Nothing, _elementTypes(i), DiagnosticInfo))
+                    End If
+
+                    members.Add(New TupleErrorFieldSymbol(Me, Name, i,
+                                                      If(_elementLocations.IsDefault, Nothing, _elementLocations(i)),
```

Comment:
This is repeated several times. Consider extracting a `GetLocation(locationsOpt, index)` helper method.